src/async_reader.rs

```diff
@@ -114,6 +114,44 @@ impl AsyncFileReader for ObjectReader {
     }
 }
 
+pub struct PrefetchReader {
+    reader: Box<dyn AsyncFileReader>,
+    buffer: Bytes,
+}
+
+impl PrefetchReader {
+    pub async fn new(mut reader: Box<dyn AsyncFileReader>, prefetch: u64) -> Result<Self> {
+        let buffer = reader.get_bytes(0..prefetch).await?;
+        Ok(Self { reader, buffer })
+    }
+}
+
+impl AsyncFileReader for PrefetchReader {
+    fn get_bytes(&mut self, range: Range<u64>) -> BoxFuture<'_, Result<Bytes>> {
+        if range.start < self.buffer.len() as _ {
+            if range.end < self.buffer.len() as _ {
+                let usize_range = range.start as usize..range.end as usize;
+                let result = self.buffer.slice(usize_range);
+                async { Ok(result) }.boxed()
```

Comment:
Then if the start and end of the requested range are fully within the existing cached buffer, we return that slice.